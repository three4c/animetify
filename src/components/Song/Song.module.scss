.Song {
  & &__list {
    margin: 0 24px;
  }

  & &__item:not(:last-child) {
    padding-bottom: 16px;
  }

  & &__title {
    margin-bottom: 2px;
  }

  & &__artist {
    line-height: 1.2;
  }

  & &__badge {
    margin-bottom: 8px;

    &Description {
      border-radius: 10px;
      color: #fff;
      display: inline-block;
      padding: 0 16px;
      line-height: 20px;
      height: 20px;
    }

    &Description {
      margin-right: 4px;
      font-weight: 500;

      &Op {
        background-color: #fcad43;

        & + .Song__badgeBroadcastOrder {
          background-color: #fcad43;
        }
      }

      &Ed {
        background-color: #0b62e5;

        & + .Song__badgeBroadcastOrder {
          background-color: #0b62e5;
        }
      }

      &Spot {
        background-color: #3cb371;
      }
    }
  }

  & &__no {
    font-size: 1.4rem;
  }

  & &__expand {
    overflow: hidden;
    position: relative;
    transition: max-height 0.4s cubic-bezier(0, 0, 0.2, 1);

    &::before {
      background: linear-gradient(
        rgba(255, 255, 255, 0),
        rgba(255, 255, 255, 1)
      );
      content: '';
      transition: height 0.4s cubic-bezier(0, 0, 0.2, 1);
      position: absolute;
      bottom: 0;
      left: 0;
      height: 0;
      width: 100%;
    }

    &Close {
      &::before {
        height: 152px;
      }
    }
  }

  & &__button {
    background-color: #fff;
    border: none;
    border-radius: 0 0 16px 16px;
    cursor: pointer;
    font-size: 14px;
    font-weight: 700;
    padding: 24px 0;
    margin-bottom: -24px;
    position: relative;
    z-index: 1;
    width: 100%;
  }
}
