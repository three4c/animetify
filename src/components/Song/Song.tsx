import React, { useState, useEffect } from 'react';
import Classnames from 'classnames';

import styles from './Song.module.scss';

interface SongProps {
  items: {
    animeDescription: string;
    animeName: string;
    artistName: string;
    broadcastOrder: string;
    musicName: string;
  }[];
}

const Song: React.FC<SongProps> = (props) => {
  const MAX_SONG_LENGTH = 3;

  const [isOpen, setOpen] = useState(false);
  const [listMinHeight, setListMinHeight] = useState(0);
  const [listMaxHeight, setListMaxHeight] = useState(0);

  const liRef = Array.from({ length: props.items.length }, () =>
    React.createRef<HTMLLIElement>()
  );

  const classNameForExpand = Classnames('Song__expand', {
    'Song__expand--close': !isOpen && props.items.length > MAX_SONG_LENGTH,
  });

  useEffect(() => {
    let minSum = 0;
    let maxSum = 0;

    liRef.forEach((_, index) => {
      if (liRef[index].current) {
        const liRefCurrent = liRef[index].current;

        if (liRefCurrent) {
          if (index < MAX_SONG_LENGTH) {
            minSum += liRefCurrent.clientHeight;
          }

          maxSum += liRefCurrent.clientHeight;
        }
      }
    });

    setListMinHeight(minSum);
    setListMaxHeight(maxSum);
  }, [liRef]);

  const buttonHandler = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    setOpen(!isOpen);
  };

  return (
    <div className={styles.Song}>
      <div
        className={classNameForExpand}
        aria-expanded={isOpen}
        style={{
          maxHeight: `${
            isOpen
              ? `${listMaxHeight}px`
              : listMinHeight
              ? `${
                  listMinHeight -
                  (props.items.length > MAX_SONG_LENGTH ? 24 : 0)
                }px`
              : 'auto'
          }`,
        }}
      >
        <ol className={styles.Song__list}>
          {props.items.map((item, index) => {
            const classNamesForDescription = Classnames(
              'Song__badgeDescription',
              {
                [`Song__badgeDescription--${item.animeDescription}`]: item.animeDescription,
              }
            );

            const classNamesForBroadcastOrder = Classnames(
              'Song__badgeBroadcastOrder',
              {
                'Song__badgeBroadcastOrder--spot':
                  item.broadcastOrder.length > 2,
              }
            );

            return (
              <li key={index} className={styles.Song__item} ref={liRef[index]}>
                <div className={styles.Song__badge}>
                  <span className={classNamesForDescription}>
                    {item.animeDescription}
                  </span>
                  {item.broadcastOrder && (
                    <span className={classNamesForBroadcastOrder}>
                      {item.broadcastOrder}
                    </span>
                  )}
                </div>
                <h3 className={styles.Song__title}>{item.musicName}</h3>
                <p className={styles.Song__artist}>{item.artistName}</p>
              </li>
            );
          })}
        </ol>
      </div>
      {props.items.length > MAX_SONG_LENGTH && (
        <button
          className={`${styles.Song__button} util-focus`}
          onClick={buttonHandler}
          aria-label="もっと見るボタン"
        >
          {isOpen ? '閉じる' : 'もっと見る'}
        </button>
      )}
    </div>
  );
};

export default React.memo(Song);
