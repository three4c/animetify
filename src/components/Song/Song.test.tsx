import React from 'react';

import { shallow, mount } from 'enzyme';

import Song from './Song';
import { act } from 'react-dom/test-utils';

describe('<Song />', () => {
  const items = [
    {
      animeName: '電波女と青春男',
      animeClassification: 'TV',
      musicName: 'Os-宇宙人',
      artistName: '大亀あすか',
      animeDescription: 'OP',
      broadcastOrder: '',
      broadcastStartDate: '2011-04-15',
    },
    {
      animeName: '電波女と青春男',
      animeClassification: 'TV',
      musicName: 'ルル',
      artistName: 'やくしまるえつこ',
      animeDescription: 'ED',
      broadcastOrder: '',
      broadcastStartDate: '2011-04-15',
    },
  ];

  it('SnapShot（works）', () => {
    expect(shallow(<Song items={items} />)).toMatchSnapshot();
  });

  //   it('useEffect Test', () => {
  //     const listMinHeight = jest.fn();
  //     const listMaxHeight = jest.fn();
  //     const useRefMock = {
  //       current: null,
  //     };

  //     jest
  //       .spyOn(React, 'useState')
  //       .mockImplementation(() => [0, listMinHeight])
  //       .mockImplementation(() => [0, listMaxHeight]);
  //     jest.spyOn(React, 'useRef').mockReturnValueOnce(useRefMock);

  //     act(() => {
  //       mount(<Song items={items} />);
  //     });

  //     expect(listMinHeight).toHaveBeenCalledTimes(1);
  //     expect(listMaxHeight).toHaveBeenCalledTimes(1);
  //   });
});
