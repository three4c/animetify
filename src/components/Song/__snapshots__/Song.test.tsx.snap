// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Song /> SnapShot（works） 1`] = `
<div
  className="Song"
>
  <div
    aria-expanded={false}
    className="Song__expand"
    style={
      Object {
        "maxHeight": "auto",
      }
    }
  >
    <ol
      className="Song__list"
    >
      <li
        className="Song__item"
        key="0"
      >
        <div
          className="Song__badge"
        >
          <span
            className="Song__badgeDescription Song__badgeDescriptionOp"
          >
            OP
          </span>
        </div>
        <h3
          className="Song__title"
        >
          Os-宇宙人
        </h3>
        <p
          className="Song__artist"
        >
          大亀あすか
        </p>
      </li>
      <li
        className="Song__item"
        key="1"
      >
        <div
          className="Song__badge"
        >
          <span
            className="Song__badgeDescription Song__badgeDescriptionEd"
          >
            ED
          </span>
        </div>
        <h3
          className="Song__title"
        >
          ルル
        </h3>
        <p
          className="Song__artist"
        >
          やくしまるえつこ
        </p>
      </li>
    </ol>
  </div>
</div>
`;
